### ISSUE TITLE
Compiler crash (Debug Failure. False Expression: should always get typenode)

### ISSUE THREAD

**Issue by matthewerwin** [Microsoft/TypeScript#17365]
**TypeScript Version:**  2.4.0  as well as nightly (2.5.0-dev.20170719)

**Code**

Create a function with the following incorrect signature:
```public asOffsetPaging(records, totalColumns:[] = null) { }```

**Expected behavior:**
Should indicate that the totalColumns:[]=null is an incorrectly defined default parameter as well as provide information about what line/file caused the problem.

**Actual behavior:**
The compiler crashes with "Debug Failure. False expression: should always get typenode" and the stack trace is useless in finding the offending line of code.

**Comment by RyanCavanaugh** [comment]
This repro isn't sufficient to reproduce the problem. Can you provide a complete sample?

**Comment by matthewerwin** [comment]
@RyanCavanaugh that is a complete sample. I can reproduce it by pasting the following into any .ts file:

public x(t:[] = null) { }

Here is the output from  "tsc --version":

Version 2.5.0-dev.20170719

Nightly build from the 19th.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.

**Comment by matthewerwin** [comment]
@mhegazy looks like it's been fixed in latest nightly builds. Still happens in 2.4.2

In ^2.6.0 I get "Type 'null' is not assignable to type '[]'." (error TS2322) and "A tuple type element list cannot be empty." (error TS1122). Dunno if you guys care to patch in 2.4.x or 2.5.x but that's what I filed the bug against.

**Comment by ktalebian** [comment]
@matthewerwin I was having the same problem when using `[]` without a type. Removing it, or creating a type for it fixed my problem. i.e.:

```ts
// Before
const dropdownItems: [] = [
    {
        value: 'id1',
        label: 'One',
    },
    {
        ...
    }
   ...
]

// After

// Either
const dropdownItems = [...]

// OR
type DropDown = {
    value: string;
    label: string;
}
const dropdownItems: DropDown[] = [...]
```

**Comment by matthewerwin** [comment]
@ktalebian thanks -- yeah, it was def some bad code but didn't expect the compiler to crash over it. We resolved it internally as well just didn't want others to run into the issue unnecessarily. Harder to know what caused it when you convert an entire project over from JS to TS & the compiler is crashing :-)

**Comment by skiabox** [comment]
This is still happening with typescript v2.4.2
I had to remove [] as a type declaration and replace it with a IBeer[] (IBeer is an interface).

**Comment by ikatyang** [comment]
Fixed by #17311 (v2.5)
